docs/adr/0006-process-for-onboarding-new-tenants.md

```diff
@@ -0,0 +1,299 @@
+
+## Decision
+
+We are envisaging a central control-plane API (referred to for the remainder of this document as the "Management API") for Marain which primarily builds on top of the Tenancy service. This will provide the standard operations such as creating new tenants and enrolling them to use Marain services.
+
+It will also need to allow us to manage licencing, but we anticipate that this will be via an as-yet nonexistent API, as it will be a central service much like tenancy, used by all other services (either directly or via something like APIM).
+
+### Onboarding
+
+Onboarding is a relatively simple part of the process where we create a new Tenant for the client. We will need to determine how we intend licencing to work and what part, if any, the Management API plays..
+
+### Enrollment
+
+Service enrollment is a more interesting aspect of the process. In order to avoid tightly coupling Marain services to the Management API, we need two things:
+- A means of discovering the available services.
+- A means of requesting that a service enroll a Tenant to use it.
+
+As described in ADR 0005, we are envisaging that each service has a Tenant created for it, under a single parent for all Service Tenants. These tenants can then underpin the discovery mechanism that allows the management API to enumerate services that tenants can be enrolled into. Part of the data attached to each tenant will be the URIs at which the endpoints they expose can be found.
+
+Once we have provided a discovery mechanism, we need to define a way in which a service can be asked to enroll a tenant. The simplest way for this to work is to require all Marain services to have a control plane API that implements a set of standard endpoints. We can provide a standard OpenAPI document that defines the endpoints needed, which we would currently expect to be:
+- Request the configuration that is needed to enroll a tenant to the service
+- Enroll a tenant to the service
+- Unenroll a tenant from the service
+
+The idea being that once the Management API has discovered a service, it can request the information that would be required to enroll a tenant to it. This will be a list of required configuration settings, in a known form that allows them to be easily captured and validated before passing them to the enrollment endpoint.
+
+The enrollment endpoint for a service will do two things:
+- Firstly, it will attach the relevant configuration to the tenant that's being enrolled.
+- Secondly, if the service that's being enrolled in has dependencies on other services, it will create a new sub-tenant of its own Service Tenant that will be used to access those dependent services. This new subtenant will then be enrolled for the dependent service.
```

Comment:
Again, If you broadcast a pub/sub message, any dependent service could also broadcast a message for its dependencies (and so on, and so forth).